src/group_theory/group_ring.lean

```diff
@@ -0,0 +1,227 @@
+
+## Tags
+
+group ring, group cohomology, monoid algebra
+-/
+
+noncomputable theory
+
+/-- The group ring `k[G]` of a group `G` with coefficients in a ring `k` (although this is defined
+  for any type `G` and any semiring `k`, inducing less algebraic structure). -/
+abbreviation group_ring (k : Type*) [semiring k] (G : Type*) := monoid_algebra k G
+
+namespace group_ring
+
+open monoid_algebra
+
+/-- Given a map `• : G × M → M` for `M` a `k`-module, the map `k[G] × M → M` sending
+`(Σ kᵢgᵢ, m) ↦ Σ kᵢ• (gᵢ • m)`. If this were an instance, it would conflict with the natural
+action of `k[k]` on itself given by multiplication. -/
+protected def has_scalar {k : Type*} [semiring k] {G : Type*} {M : Type*}
+  [add_comm_monoid M] [has_scalar G M] [module k M] :
+  has_scalar (group_ring k G) M :=
+⟨λ g m, finsupp.total G M k (λ g, g • m) g⟩
+
+section
+
+local attribute [instance] group_ring.has_scalar
+
+lemma smul_def {k : Type*} [semiring k] {G : Type*} {M : Type*}
+  [add_comm_monoid M] [has_scalar G M] [module k M]
+  (r : group_ring k G) (x : M) :
+  r • x = finsupp.total G M k (λ g, g • x) r := rfl
+
+/-- Given a map `• : G × M → M` for `M` a `k`-module, `(r • Σ kᵢgᵢ) • m = r • (Σ kᵢ • (gᵢ • m))`
+for all `r ∈ k`, `Σ kᵢgᵢ ∈ k[G]`, `m ∈ M.` -/
+instance {k : Type*} [semiring k] {G : Type*}
+  {M : Type*} [add_comm_monoid M] [has_scalar G M] [module k M] :
+  is_scalar_tower k (group_ring k G) M :=
+{ smul_assoc := λ g x m, show finsupp.total _ _ _ _ _ = g • finsupp.total _ _ _ _ _,
+    by simp only [ring_hom.id_apply, linear_map.map_smulₛₗ] }
+
+/-- If `M` has compatible `G`-module and `k`-module structures for a semiring `k`,
+it is also a `k[G]`-module. -/
+def action_to_module {k : Type*} [semiring k] {G : Type*} [monoid G]
+  {M : Type*} [add_comm_monoid M] [distrib_mul_action G M] [module k M]
+  [smul_comm_class k G M] : module (group_ring k G) M :=
+{ smul := (•),
+  one_smul := λ b, by
+  { dsimp,
+    show finsupp.total _ _ _ _ _ = _,
+    erw [finsupp.total_single],
+    simp only [one_smul] },
+  mul_smul := λ g h m, by
+  { refine g.induction_on _ _ _,
+    { intros g,
+      dsimp,
+      rw [monoid_algebra.mul_def, finsupp.sum_single_index],
+      { simp only [smul_def, one_mul, one_smul, finsupp.total_single],
+        simp only [finsupp.total_single, linear_map.map_finsupp_sum],
+        rw [finsupp.total_apply, finsupp.smul_sum],
+        simp only [mul_smul, smul_comm] },
+      { simp }},
+    { intros x y hx hy,
+      simp only [smul_def, add_mul, linear_map.map_add] at ⊢ hx hy,
+      simp only [hx, hy] },
+    { intros r f hf,
+      rw [smul_assoc, smul_mul_assoc, smul_assoc, hf] }},
+  smul_add := λ g b c, by
+    { simp only [smul_def, smul_add, finsupp.total_apply, finsupp.sum_add], },
+  smul_zero := λ x, by
+    simp only [smul_def, finsupp.total_apply, finsupp.sum_zero, smul_zero],
+  add_smul := λ r s x, linear_map.map_add _ _ _,
+  zero_smul := λ x, linear_map.map_zero _ }
```

Comment:
This is a slightly shorter proof:
```suggestion
lemma smul_eq_lift_nc_ring_hom {k : Type*} [semiring k] {G : Type*} [monoid G] {M : Type*}
  [add_comm_monoid M] [distrib_mul_action G M] [module k M] [smul_comm_class k G M]
  (r : group_ring k G) (x : M) :
  r • x = lift_nc_ring_hom (module.to_add_monoid_End k M) (distrib_mul_action.to_add_monoid_End _ M)
      (λ _ _, add_monoid_hom.ext $ smul_comm _ _) r x :=
eq.trans (by refl) (add_monoid_hom.finsupp_sum_apply r _ x : _).symm

/-- If `M` has compatible `G`-module and `k`-module structures for a semiring `k`,
it is also a `k[G]`-module. -/
def action_to_module {k : Type*} [semiring k] {G : Type*} [monoid G]
  {M : Type*} [add_comm_monoid M] [distrib_mul_action G M] [module k M]
  [smul_comm_class k G M] : module (group_ring k G) M :=
{ smul := (•),
  one_smul := λ b, by rw [action_eq, map_one, add_monoid.coe_one, id],
  mul_smul := λ g h m, by rw [action_eq, action_eq, action_eq, map_mul, add_monoid.coe_mul],
  smul_add := λ g b c, by rw [action_eq, action_eq, action_eq, add_monoid_hom.map_add],
  smul_zero := λ x, by rw [action_eq, add_monoid_hom.map_zero],
  add_smul := λ r s x, linear_map.map_add _ _ _,
  zero_smul := λ x, linear_map.map_zero _ }

```

Reply:
thank you for this but I'm trying to implement it (having replaced `action_eq` with `smul_eq_lift_nc_ring_hom`) and although it compiles, my later stuff (e.g. when I use `action_to_module` on line 193) breaks... Guess Lean is proof relevant now haha

Reply:
For clarity, I note you are right that we need to replace `action_eq` with `smul_eq_lift_nc_ring_hom` for this suggestion.

The problems which you mention this triggers later in the file seem to be caused by some odd universe inference logic in Lean. For example if instead of:
```lean
lemma smul_eq_lift_nc_ring_hom {k : Type*} [semiring k] {G : Type*} [monoid G] {M : Type*}
```
we write:
```lean
universes u₁ u₂ u₃

lemma smul_eq_lift_nc_ring_hom {k : Type u₁} [semiring k] {G : Type u₂} [monoid G] {M : Type u₃}
```
then all is well!

I suggest adding the `universes` statement at the top of the file, using Eric's suggestion with this modification and asking on Zulip once these changes have passed CI.